File: 205.同构字符串.js
/*
 * @lc app=leetcode.cn id=205 lang=javascript
 *
 * [205] 同构字符串
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isIsomorphic = function(s, t) {
    let s2t = new Map();
    let t2s = new Set();

    for (let i = 0; i < s.length; i++) {
        // 检查是否有多对一
        if (!s2t.has(s.charAt(i))) {
            if (t2s.has(t.charAt(i))) {
                return false;
            }
            s2t.set(s.charAt(i), t.charAt(i));
            t2s.add(t.charAt(i));
        // 检查是否有一对多
        }else if (t.charAt(i) != s2t.get(s.charAt(i))) {
            return false
        }
    }
    return true;
};
// @lc code=end

File: 205.同构字符串.go
/*
 * @lc app=leetcode.cn id=205 lang=golang
 *
 * [205] 同构字符串
 */

// @lc code=start
func isIsomorphic(s string, t string) bool {
	sr, tr := []rune(s), []rune(t)
	s2t := make(map[rune]rune)
	tset := make(map[rune]struct{})
	for i, sb := range sr {
		if _, ok := s2t[sb]; !ok {
			if _, ok1 := tset[tr[i]]; ok1 {
				return false
			}

			s2t[sb] = tr[i]
			tset[tr[i]] = struct{}{}
		} else {
			if s2t[sb] != tr[i] {
				return false
			}
		}
	}

	return true
}

// @lc code=end

